src/GobanCore.ts

```diff
@@ -2153,14 +2160,54 @@ export abstract class GobanCore extends TypedEventEmitter<Events> {
 
         /* This must be done last as it will invoke the appropriate .set actions to set the board in it's correct state */
         const old_engine = this.engine;
-        this.engine = new GoEngine(config, this);
-        this.engine.parentEventEmitter = this;
-        this.engine.getState_callback = () => {
-            return this.getState();
-        };
-        this.engine.setState_callback = (state) => {
-            return this.setState(state);
-        };
+
+        // we need to have an engine to be able to keep it
+        keep_old_engine = keep_old_engine && old_engine !== null && old_engine !== undefined;
+        // we only keep the old engine in analyze mode & finished state
+        // JM: this keep_old_engine functionality is being added to fix resetting analyze state on network
+        // reconnect
+        keep_old_engine =
+            keep_old_engine && this.mode === "analyze" && old_engine.phase === "finished";
```

Comment:
0) safety, with my (lame) understanding of what client <-> server interactions are going on during live game (other specifics of demo/rengo/puzzles/whatnot), I would not dare

1) I do not see the upside for play, I thought that you need to submit move anyway, so "server is always right", relaxing this & keeping the local state could possibly introduce some nasty bugs

2) assuming it would actually make sense, much more of "is the information in new contained in old" checks would be needed imo; the `MoveTree` has much more state than what is actually checked here (I really just check the trees' `(x,y,player) triple`) & much more than I understand - marks, chatlog, ... & `GoEngine` as well. I do not understand the internals/interactions (e.g. what all things could be contained in gamedata sent from server, ...), so I would not dare making such a change. I view this change as a reasonably safe fix, that could be a bit broken (e.g. the marks maybe ?..) but only locally in analyze, and much less than is broken now, so enabling for analyze is a clear win in my view. To use it in play, I think the possibility of introducing bugs is not worth the limited upside as per 1) (but maybe the upside is bigger, dunno). Also this change potentially makes the code less future-bugs-proof - as changing/adding some state that should be checked without adding the checks (on when to change/keep engine) could again cause bugs.

Imo not worth the work? But maybe I am just overly cautious here, hard for me to judge. Wdyt @anoek ?